tests/consumer_tests/routers/test_ucam.py

```diff
@@ -0,0 +1,148 @@
+
+        result = client.get("/ucam/patients/E-PATIENT").json()
+
+        assert result["meta"]["success"] is True
+        assert len(result["data"]) > 0
+
+
+def test_get_patient_not_found(mock_data: dict, client: TestClient) -> None:
+    with patch.object(ucam, "response", return_value=None):
+
+        result = client.get("/ucam/patients/NO-PATIENT").json()
+
+        assert len(result["meta"]["errors"]) > 0
+        assert result["meta"]["success"] is False
+        assert result["data"] is None
+
+
+def test_get_patients_success(mock_data: dict, client: TestClient) -> None:
+    with patch.object(ucam, "response", return_value=mock_data["patients"]):
+
+        result = client.get("/ucam/patients/").json()
+
+        assert len(result["data"]) == 6
+        assert result["meta"]["success"] is True
+
+
+def test_get_device_success(mock_data: dict, client: TestClient) -> None:
+    with patch.object(
+        ucam,
+        "response",
+        return_value=[
+            d for d in mock_data["devices"] if d["device_id"] == "NR3-DEVICE"
+        ],
+    ):
+
+        result = client.get("/ucam/devices/NR3-DEVICE").json()
+
+        assert result["meta"]["success"] is True
+        assert len(result["data"]) == 1
+        assert result["data"][0]["device_id"] == "NR3-DEVICE"
+
+
+def test_get_device_not_found(mock_data: dict, client: TestClient) -> None:
+    with patch.object(ucam, "response", return_value=None):
+
+        result = client.get("/ucam/devices/NOT-DEVICE").json()
+
+        assert result["meta"]["success"] is False
+        assert len(result["meta"]["errors"]) > 0
+        assert result["data"] is None
+
+
+def test_get_devices_success(mock_data: dict, client: TestClient) -> None:
+    with patch.object(ucam, "response", return_value=mock_data["devices"]):
+
+        result = client.get("/ucam/devices/").json()
+
+        assert len(result["data"]) == 5
+        assert result["meta"]["success"] is True
+
+
+def test_get_vtt_success(mock_data: dict, client: TestClient) -> None:
+    with patch.object(ucam, "response", return_value=mock_data["vtt"]):
+
+        result = client.get("/ucam/vtt/VTT_COMPLEX_HASH").json()
+
+        assert len(result["data"]) == 2
+        assert result["meta"]["success"] is True
+
+
+def test_get_vtts_success(mock_data: dict, client: TestClient) -> None:
+    with patch.object(ucam, "response", return_value=mock_data["vtt"]):
+
+        result = client.get("/ucam/vtt/").json()
+
+        assert len(result["data"]) == 2
+        assert result["meta"]["success"] is True
+
+
+@pytest.mark.live
+def test_LIVE_auth_OK() -> None:
+    with patch.dict(
+        os.environ, {"UCAM_ACCESS_TOKEN": "", "UCAM_ACCESS_TOKEN_GEN_TIME": "0"}
+    ):
+
+        result = ucam.__ucam_access_token()
+
+        assert isinstance(result, str)
+
+
+@pytest.mark.live
+def test_LIVE_get_patients_OK() -> None:
+
+    result = ucam.get_patients()
+
+    assert all(isinstance(x, Patient) for x in result)
+
```

Comment:
I've worked on this a bit, but for some reason could not get pytest to pick up on the 'functional' folder. Also - as we discussed quite a while a go - I can't name the test subfolder the same name as the module folder, as then any module imports will fail.

In addition, I was refactoring a very large number of things here, and since this branch is not up to date with master (and overall this is getting really big now), I'd like to move this into a different issue and PR.